src/cast_core/src/SampleHttpClient.cpp

```diff
@@ -0,0 +1,129 @@
+};
+
+
+//
+// Genny spins up a thread for each Actor instance. The `Threads:` configuration
+// tells Genny how many such instances to create. See further documentation in
+// the `Actor.hpp` file.
+//
+void SampleHttpClient::run() {
+    for (auto&& config : _loop) {
+        for (const auto&& _ : config) {
+            auto document = config->documentExpr();
+
+            auto requests = _totalRequests.start();
+
+            BOOST_LOG_TRIVIAL(debug) << " SampleHttpClient Inserting "
+                                    << bsoncxx::to_json(document.view());
+
+            try {
+                boost::asio::io_context ioContext;
+                auto client = std::make_shared<simple_http::get_client>(
+                    ioContext,
+                    [](simple_http::empty_body_request& /*req*/,
+                    simple_http::string_body_response& resp) {
+                    // noop for successful HTTP
+                    });
+
+                client->setFailHandler([&requests](const simple_http::empty_body_request& /*req*/,
+                            const simple_http::string_body_response& /*resp*/,
+                            simple_http::fail_reason fr, boost::string_view message) {
+                        std::cout << "Failure code: " << fr << '\n';
+                    requests.failure();
+                });
+
+                // Run the GET request to httpbin.org
+                client->get(simple_http::url{
+                    "https://user:passwd@httpbin.org/digest-auth/auth/user/passwd/MD5/never"});
+
+                ioContext.run();
+
+                requests.success();
+            } catch(mongocxx::operation_exception& e) {
+                requests.failure();
+                //
+                // MongoException lets you include a "causing" bson document in the
+                // exception message for help debugging.
+                //
+                BOOST_THROW_EXCEPTION(MongoException(e, document.view()));
+            } catch(...) {
+                requests.failure();
+                throw std::current_exception();
+            }
+        }
+    }
+}
+
+SampleHttpClient::SampleHttpClient(genny::ActorContext& context)
+    : Actor{context},
+      _totalRequests{context.operation("Insert", SampleHttpClient::id())},
+      _client{context.client()},
+      //
+      // Pass any additional constructor parameters that your `PhaseConfig` needs.
+      //
+      // The first argument passed in here is actually the `ActorContext` but the
+      // `PhaseLoop` reads the `PhaseContext`s from there and constructs one
+      // instance for each Phase.
+      //
+      _loop{context, (*_client)[context["Database"].to<std::string>()], SampleHttpClient::id()} {}
+
+namespace {
+//
+// This tells the "global" cast about our actor using the defaultName() method
+// in the header file.
+// Do not use simple-beast-client on macOS
+#ifndef __APPLE__
```

Comment:
I need to push back here.

1. We've jumped through enormous hoops to make all of Genny cross-platform and easy to develop on the standard-issue macbooks provided by the company plus a number of other reasonable developer platforms like arch and Ubuntu although we don't run those in production.
5. Our ["build and install" page](https://github.com/mongodb/genny/blob/6a2a0648f36b02b4121d19789346a7a1ebba110a/docs/setup.md#build-and-install) says we support macOS. (It also says we support Windows which may in fact be true since we're pretty by-the-book with our code.)
2. The toolchain is built for macOS and other such platforms at considerable development and CI build cost.
3. The CMake is standard enough to be used by all manner of IDEs
4. We only use standards-compliant C++17 and don't use compiler-specific `#pragma`s for include guards.
6. We have explicit recommendations to avoid `#ifdef` in the [contributions guide](https://github.com/mongodb/genny/blob/master/CONTRIBUTING.md#dont-get-cute).
7. The reason why this code doesn't work on macOS isn't well-understood according to my reading of this situation. Does this extend to the other platforms that we support as well? This means there could be a latent bug in the code (or the libraries we're using) that could surface at some unknown time on some arbitrary platform in the future.  I.e., are we relying on undefined or accidental behavior of linux? Or is it a "leak" in the toolchain that causes us to link to a system library that behaves differently on macOSand/or other systems?

This ["breaks the first window"](https://www.psychologytoday.com/us/basics/broken-windows-theory) of the consistent cross-platform nature of the product.

Our "definition of done" for Genny work includes ensuring that all features work across all the platforms we support. If this weren't the case we'd have abandoned macOS, Arch, RHEL, and ubuntu long ago since our actual perf tests only need to run on Amazon-Linux... (None of this is explicit anywhere so this mis-reading of the situation is totally understandable. I'll cut a ticket to ensure our docs require x-platform).

Of course this must be balanced with the need to wrap up this work and move on to other things.

May I suggest:

1. Time-boxing a debugging session with a macOS debugger to chase down what the underlying issue actually is. I had to do this with the Catch2 integration and would be happy to provide a hand or a shoulder to cry on. If we've given sufficient effort into debugging we can provide additional context and justification in the code so that someone who has more time/energy in the future can pick up the work if/when it's deemed necessary.
2. Explicitly not supporting DNS and requiring URLs to be IP based. This may not work for the endpoints in question, but it does explicitly exclude known edge-cases.
3. Reworking the way we invoke DNS to require users to provide the DNS mappings in some manner. (I have no idea how feasible this is, but ideally the internals would handle DNS via vtables which we could hook into.)

Other options exist too of course. Not trying to be unreasonable here. I'm happy to discuss further on slack/zoom.

Reply:
Hey Ryan, I'm going to continue the conversation around macOS support as part of https://jira.mongodb.org/browse/TIG-3843 and resolve this thread. The code currently builds on macOS and anyone not developing this actor would not be affected. Richard mentioned the IFDEF has been removed as well.